### ISSUE TITLE
[BUG] DefaultAzureCredential fails in production on Azure VM

### ISSUE THREAD

**Issue by jaschrep-msft** [Azure/azure-sdk-for-net#9588]
**Describe the bug**
`DefaultAzureCredential` fails to find the managed identity endpoint in a production build on an Azure VM (there is a rare chance it succeeds). The unchanged code does not fail when debugging in Visual Studio on the exact same VM.

On my dev machine, `DefaultAzureCredential` will successfully use an `EnvironmentCredential` instead of `ManagedIdentityCredential`. As a temporary workaround, I replicated the same environment variables on the target VM, but `DefaultAzureCredential` *could not find those environment variables either* (I set them as system variables instead of user variables on the target VM to ensure Azure's Compute extensions for remote powershell scripts will have access to them).

I happen to know that MSI is functioning properly on the target VM because AzCopy manages to successfully authenticate using MSI earlier in the exact same script.

***Exception or Stack Trace***
Logs as recorded by `AzureEventSourceListener`
```
Azure.Identity.DefaultAcureCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Azure.Identity.EnvironmentCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Azure.Identity.EnvironmentCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.CredentialUnavailableException (0x80131500): Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ]
Azure.Identity.ManagedIdentityCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Request [ec635421-038f-474a-98b5-e17014f36494] GET http://169.254.169.254/metadata/identity/oauth2/token??pi-version=REDACTED 
x-ms-client-request-id:ec635421-038f-474a-98b5-e17014f36494
x-ms-return-client-request-id:true
User-Agent:azsdk-net-Identity/1.0.0+809e5374b6b040adcb6425f9e46b19f5182b3488,(.NET Core 4.6.28008.02; Microsoft Windows 10.0.17763 )

Azure.Identity.ManagedIdentityCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.CredentialUnavailableException (0x80131500): No managed identity endpoint found.
Azure.Identity.SharedTokenCacheCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Azure.Identity.SharedTokenCacheCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.CredentialUnavailableException (0x80131500): No accounts were discovered in the shared token cache. To fix, authenticate through tooling supporting azure developer sign on.
Azure.Identity.DefaultAcureCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.AuthenticationFailedException (0x80131500): The DefaultAzureCredential failed to retrieve a token from the included credentials.
  EnvironmentCredential is unavailable Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ].
  ManagedIdentityCredential is unavailable No managed identity endpoint found..
  SharedTokenCacheCredential is unavailable No accounts were discovered in the shared token cache. To fix, authenticate through tooling supporting azure developer sign on..
See inner exception for more detail.
 ---> System.AggregateException (0x80131500): The DefaultAzureCredential failed to retrieve a token from the included credentials. (Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ]) (No managed identity endpoint found.) (No accounts were discovered in the shared token cache. To fix, authenticate through tooling supporting azure developer sign on.)
 ---> Azure.Identity.CredentialUnavailableException (0x80131500): Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ]
```

**To Reproduce**
I do not have a succinct, guaranteed reproduction at this time. This error occurs in release builds run on Azure VMs, and these builds are being invoked through Azure Compute's extensions to run powershell commands on Azure VMs.

***Code Snippet***
```csharp
var serviceClient = new BlobServiceClient(
    myStorageEndpoint,
    new Azure.Identity.DefaultAzureCredential());
var containerClient = serviceClient.GetBlobContainerClient(myNewContainerName);
containerClient.Create(); //throws
```

**Expected behavior**
DefaultAzureCredential should successfully obtain a token for my requests.

**Screenshots**
N/A

**Setup:**
 - OS: Windows Datacenter 2019
 - IDE : Release build only (occasionally crops up in Visual Studio 2017)
 - Version of the Library used: Identity 1.0.0; Storage 12.0.0

**Additional context**
N/A

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by tg-msft** [comment]
The scope name also appears to have [a small typo](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/src/DefaultAzureCredential.cs#L95).

**Comment by schaabs** [comment]
@jaschrep-msft thanks for reporting this issue. I believe this is a timing issue in MSI endpoint discovery. I'm currently working on a fix in this area, and while working on that I discovered the timeout we were using to probe for the endpoint was not quite sufficient for all environments. I have adjusted this as part of that fix. I'm currently updating the PR for it and I'll update this issue once I have the fix available to test out.

**Comment by schaabs** [comment]
@tg-msft thanks for the catch on the typo, I have a fix here, https://github.com/Azure/azure-sdk-for-net/pull/9600. PTAL

**Comment by schaabs** [comment]
@jaschrep-msft There are a couple of changes in merged in PR #9025 which should address your issue. First of all the IMDS instance discovery timeout has been increased from 500ms to 1000ms. Also, the discovery is no longer cached statically. So if the original discovery fails you can retry by creating a new instance of the ManagedIdentityCredential.

This fix will be included in the next release of Azure.Identity which is currently scheduled for mid February. Please feel free to reopen this issue if you still have problems with this fix in place.
